AK/SipHash.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/MemoryStream.h>
+#include <AK/SipHash.h>
+#include <AK/Span.h>
+#include <AK/UFixedBigInt.h>
+
+#ifdef KERNEL
+#    include <Kernel/Security/Random.h>
+#else
+#    include <AK/Random.h>
+#endif
+
+namespace AK {
+
+#define ROTL(x, b) (u64)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define SIPROUND           \
+    do {                   \
+        v0 += v1;          \
+        v1 = ROTL(v1, 13); \
+        v1 ^= v0;          \
+        v0 = ROTL(v0, 32); \
+        v2 += v3;          \
+        v3 = ROTL(v3, 16); \
+        v3 ^= v2;          \
+        v0 += v3;          \
+        v3 = ROTL(v3, 21); \
+        v3 ^= v0;          \
+        v2 += v1;          \
+        v1 = ROTL(v1, 17); \
+        v1 ^= v2;          \
+        v2 = ROTL(v2, 32); \
+    } while (0)
+
+template<size_t message_block_rounds, size_t finalization_rounds>
+static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
+{
+    VERIFY((output.size() == 8) || (output.size() == 16));
+
+    u64 v0 = 0x736f6d6570736575ull;
+    u64 v1 = 0x646f72616e646f6dull;
+    u64 v2 = 0x6c7967656e657261ull;
+    u64 v3 = 0x7465646279746573ull;
+    auto const left = input.size() & 7;
+    // The end of 64-bit blocks.
+    auto const block_end = input.size() - (input.size() % sizeof(u64));
+    u64 b = input.size() << 56;
+    v3 ^= key.high();
+    v2 ^= key.low();
+    v1 ^= key.high();
+    v0 ^= key.low();
+
+    if (output.size() == 16)
+        v1 ^= 0xee;
+
+    FixedMemoryStream input_stream { input };
```

Comment:
This right here is holding your perf back by at least 30%.

Reply:
Without streams all of the endianness conversion will look extremely ugly and involve pointer casts everywhere, is that better?

Reply:
```diff
--- a/AK/SipHash.cpp
+++ b/AK/SipHash.cpp
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/ByteReader.h>
 #include <AK/MemoryStream.h>
 #include <AK/SipHash.h>
 #include <AK/Span.h>
@@ -58,10 +59,9 @@ static void do_siphash(ReadonlyBytes input, u128 key, Bytes output)
     if (output.size() == 16)
         v1 ^= 0xee;
 
-    FixedMemoryStream input_stream { input };
     for (size_t input_index = 0; input_index < block_end; input_index += 8) {
         // The way we use them, no stream functions can fail.
-        auto const m = MUST(input_stream.read_value<LittleEndian<u64>>());
+        auto const m = bit_cast<LittleEndian<u64>>(ByteReader::load64(input.slice(input_index, sizeof(u64)).data()));
         v3 ^= m;
 
         for (size_t i = 0; i < message_block_rounds; ++i)
```

:yakfused: ugly?